### ISSUE TITLE
create a new layer with mx.operator.CustomOpProp in mxnet 

### ISSUE THREAD

**Issue by xzqjack** [apache/incubator-mxnet#4205]
Hi. I wanna create a layer **perturbation** that generate a noisy tensor **N** (loc=1, scale =1) to perturb input tensor **Input**. The **output**  is **N*Input**. But some error about "asynchronous engine operation" occured. Below is the some information. Any hits will be appreciated.

## Environment info
Operating System: Ubuntu 16.04, CPU

Compiler: gcc 5.4.0

Package used (Python/R/Scala/Julia): Python 2.7

MXNet version: 0.7.0

## Error Message:

> runfile('/home/xzqjack/mxnet/example/numpy-ops/permutation_layer.py', wdir='/home/xzqjack/mxnet/example/numpy-ops')
> [22:26:08] src/operator/./softmax_output-inl.h:329: Softmax symbol is renamed to SoftmaxOutput. This API will be deprecated in Dec, 2015
> [22:26:09] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(100,784)
> [22:26:10] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(100,784)
> INFO:root:Start training with [cpu(0)]
> ('Error in CustomOp.forward: ', "object of type 'long' has no len()")
> [22:26:10] /home/xzqjack/mxnet/dmlc-core/include/dmlc/logging.h:235: [22:26:10] src/operator/custom.cc:79: Check failed: op_info_->forward(ptrs.size(), ptrs.data(), tags.data(), reqs.data(), ctx.is_train, op_info_->p_forward) 
> [22:26:10] /home/xzqjack/mxnet/dmlc-core/include/dmlc/logging.h:235: [22:26:10] src/engine/./threaded_engine.h:306: [22:26:10] src/operator/custom.cc:79: Check failed: op_info_->forward(ptrs.size(), ptrs.data(), tags.data(), reqs.data(), ctx.is_train, op_info_->p_forward) 
> An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPE to NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
> terminate called after throwing an instance of 'dmlc::Error'
>   what():  [22:26:10] src/engine/./threaded_engine.h:306: [22:26:10] src/operator/custom.cc:79: Check failed: op_info_->forward(ptrs.size(), ptrs.data(), tags.data(), reqs.data(), ctx.is_train, op_info_->p_forward) 
> An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPE to NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
> 

## Minimum reproducible example

```
# pylint: skip-file
import os
# MXNET_CPU_WORKER_NTHREADS must be greater than 1 for custom op to work on CPU
os.environ["MXNET_CPU_WORKER_NTHREADS"] = "2"
from data import mnist_iterator
import mxnet as mx
import logging

class perturbation(mx.operator.CustomOp):
    def __init__(self, ctx):
        super(perturbation, self).__init__()       
        self.rand = None 
        self.ctx = ctx
        
    def forward(self, is_train, req, in_data, out_data, aux):
        x = in_data[0]
        self.rand = mx.random.normal(loc=1, scale=1,shape=x.shape[0], ctx=self.ctx)
        y = x*self.rand
        self.assign(out_data[0], req[0], y)

    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        y = self.rand*out_grad
        self.assign(in_grad[0], req[0], y)
        
@mx.operator.register("perturbation")
class perturbationProp(mx.operator.CustomOpProp):
    def __init__(self):
        super(perturbationProp, self).__init__(need_top_grad=True)
    
    def list_arguments(self):
        return ['data']

    def list_outputs(self):
        return ['output']
        
    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        output_shape = in_shape[0]
        return [data_shape], [output_shape], []

    def create_operator(self, ctx, shapes, dtypes):
        return perturbation(ctx)

# define mlp
data = mx.symbol.Variable('data')

##This is the new defined layer
fc1 = mx.symbol.FullyConnected(data = data, name='fc1', num_hidden=128)

fc1 = mx.symbol.Custom(data=fc1, name='perturbation', op_type='perturbation')
act1 = mx.symbol.Activation(data = fc1, name='relu1', act_type="relu")
fc2 = mx.symbol.FullyConnected(data = act1, name = 'fc2', num_hidden = 64)
act2 = mx.symbol.Activation(data = fc2, name='relu2', act_type="relu")
fc3 = mx.symbol.FullyConnected(data = act2, name='fc3', num_hidden=10)
mlp = mx.symbol.Softmax(data = fc3, name = 'softmax')

# data

train, val = mnist_iterator(batch_size=100, input_shape = (784,))

# train

logging.basicConfig(level=logging.DEBUG)

# MXNET_CPU_WORKER_NTHREADS must be greater than 1 for custom op to work on CPU
model = mx.model.FeedForward(
    ctx = mx.cpu(0), symbol = mlp, num_epoch = 10,
    learning_rate = 0.1, momentum = 0.9, wd = 0.00001)

model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(100,100))

```

## Steps to reproduce
1. The code is modified from mxnet/example/numpy-ops/custom_softmax.py.  If you want run, you should download mnist dataset first.

## What have you tried to solve it?
I created the new layer by referencing "mxnet/example/numpy-ops/custom_softmax.py" and some other example codes.

**Comment by kevinthesun** [comment]
Some errors:
1. in "self.rand = mx.random.normal(loc=1, scale=1,shape=x.shape[0], ctx=self.ctx)", shape should be a tuple. I changed it to "self.rand = mx.random.normal(loc=1, scale=1, shape=(x.shape[0], 1), ctx=self.ctx)"

2. in "self.ctx = ctx", cox should be a mx.Context object, but mx.cpu(0) will return a str of device id. I changed it to "self.ctx = mx.Context(ctx)"

3. in "y = self.rand*out_grad", out_grad is a list of NDArray. Here I changed it to "y = self.rand * out_grad[0]"

Complete code:
`# pylint: skip-file
import os
# MXNET_CPU_WORKER_NTHREADS must be greater than 1 for custom op to work on CPU
os.environ["MXNET_CPU_WORKER_NTHREADS"] = "2"
from data import mnist_iterator
import mxnet as mx
import logging

class perturbation(mx.operator.CustomOp):
    def __init__(self, ctx):
        super(perturbation, self).__init__()       
        self.rand = None 
        self.ctx = mx.Context(ctx)
        
    def forward(self, is_train, req, in_data, out_data, aux):
        x = in_data[0]
        self.rand = mx.random.normal(loc=1, scale=1, shape=(x.shape[0], 1), ctx=self.ctx)
        y = x * self.rand
        self.assign(out_data[0], req[0], y)

    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        y = self.rand * out_grad[0]
        self.assign(in_grad[0], req[0], y)
        
@mx.operator.register("perturbation")
class perturbationProp(mx.operator.CustomOpProp):
    def __init__(self):
        super(perturbationProp, self).__init__(need_top_grad=True)
    
    def list_arguments(self):
        return ['data']

    def list_outputs(self):
        return ['output']
        
    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        output_shape = in_shape[0]
        return [data_shape], [output_shape], []

    def create_operator(self, ctx, shapes, dtypes):
        return perturbation(ctx)

# define mlp
data = mx.symbol.Variable('data')

##This is the new defined layer
fc1 = mx.symbol.FullyConnected(data = data, name='fc1', num_hidden=128)

fc1 = mx.symbol.Custom(data=fc1, name='perturbation', op_type='perturbation')
act1 = mx.symbol.Activation(data = fc1, name='relu1', act_type="relu")
fc2 = mx.symbol.FullyConnected(data = act1, name = 'fc2', num_hidden = 64)
act2 = mx.symbol.Activation(data = fc2, name='relu2', act_type="relu")
fc3 = mx.symbol.FullyConnected(data = act2, name='fc3', num_hidden=10)
mlp = mx.symbol.Softmax(data = fc3, name = 'softmax')

# data

train, val = mnist_iterator(batch_size=100, input_shape = (784,))

# train

logging.basicConfig(level=logging.DEBUG)

# MXNET_CPU_WORKER_NTHREADS must be greater than 1 for custom op to work on CPU
model = mx.model.FeedForward(
    ctx = mx.cpu(0), symbol = mlp, num_epoch = 10,
    learning_rate = 0.1, momentum = 0.9, wd = 0.00001)

model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(100,100))`

**Comment by xzqjack** [comment]
@kevinthesun  Thanks for your help. 
Your modification make the code works. There are a lot i need to learn.

**Comment by xzqjack** [comment]
Hi @kevinthesunI. 
In order to make the perturbation layer suitable to Non-FC layer too, such as conv layer, I change "shape=(x.shape[0], 1)" to " shape = (x.shape[0],1) if len(x.shape)==1 else x.shape"
. It works fine on "mxent/example/image-classification"{train_mnist.py, train_cifar10.py(model-resnet-28-small, model-inception-bn-28-small)} and accuracy increases. However, when "train_cifar10.py(model-symbol_vgg)", the running seems stuck in somewhere.

> runfile('/home/gpu3/Anchor/featureMapAugmentation/train_cifar10.py', wdir='/home/gpu3/Anchor/featureMapAugmentation')
> 2016-12-13 15:39:24,179 Node[0] start with arguments Namespace(batch_size=16, data_dir='cifar10/', gpus='0', kv_store='local', load_epoch=None, lr=0.05, lr_factor=1, lr_factor_epoch=1, model_prefix=None, network='anchor_vgg', num_epochs=20, num_examples=60000, save_model_prefix=None)
> [15:39:24] src/io/iter_image_recordio.cc:211: ImageRecordIOParser: cifar10/train.rec, use 3 threads for decoding..
> [15:39:24] src/io/./iter_normalize.h:103: Load mean image from cifar10/mean.bin
> [15:39:24] src/io/iter_image_recordio.cc:211: ImageRecordIOParser: cifar10/test.rec, use 3 threads for decoding..
> [15:39:24] src/io/./iter_normalize.h:103: Load mean image from cifar10/mean.bin
> 2016-12-13 15:39:24,907 Node[0] Start training with [gpu(0)]
> (16L, 64L, 28L, 28L)

Stuck in here

The new layer is
```
import os
# MXNET_CPU_WORKER_NTHREADS must be greater than 1 for custom op to work on CPU
os.environ["MXNET_CPU_WORKER_NTHREADS"] = "2"

class perturbation(mx.operator.CustomOp):
    def __init__(self, ctx):
        super(perturbation, self).__init__()       
        self.rand = None 
        self.ctx = mx.Context(ctx)
        
    def forward(self, is_train, req, in_data, out_data, aux):
        x = in_data[0]
        print x.shape
        shape = (x.shape[0],1) if len(x.shape)==1 else x.shape
        self.rand = mx.random.normal(loc=1, scale=0, shape=shape, ctx=self.ctx)
        y = x*self.rand
        self.assign(out_data[0], req[0], y)

    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        y = self.rand*out_grad[0]
        self.assign(in_grad[0], req[0], y)
        
@mx.operator.register("perturbation")
class perturbationProp(mx.operator.CustomOpProp):
    def __init__(self):
        super(perturbationProp, self).__init__(need_top_grad=True)
    
    def list_arguments(self):
        return ['data']

    def list_outputs(self):
        return ['output']
        
    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        output_shape = in_shape[0]
        return [data_shape], [output_shape], []

    def create_operator(self, ctx, shapes, dtypes):
        return perturbation(ctx)
```

The perturbation layer works fine on "resnet-28-small", "inception-bn-28-small", and works worse on "vgg". I have no idea what is the difference...

**Comment by kevinthesun** [comment]
Maybe add some logging statements and see what's happening?
